src/riverqueue/client.py

```diff
@@ -569,145 +534,79 @@ def insert_many_tx(self, tx, args: List[JobArgs | InsertManyParams]) -> int:
-            return InsertResult(exec.job_insert(insert_params))
+        return self._insert_many_exec(self.driver.unwrap_executor(tx), args)
 
-        # fast path
-        if (
-            not unique_opts.by_state
-            or unique_opts.by_state.sort == UNIQUE_STATES_DEFAULT
-        ):
-            job, unique_skipped_as_duplicate = exec.job_insert_unique(
-                insert_params, sha256(unique_key.encode("utf-8")).digest()
-            )
-            return InsertResult(
-                job=job, unique_skipped_as_duplicated=unique_skipped_as_duplicate
-            )
-
-        with exec.transaction():
-            lock_key = "unique_key"
-            lock_key += "kind=#{insert_params.kind}"
-            lock_key += unique_key
-
-            exec.advisory_lock(_hash_lock_key(self.advisory_lock_prefix, lock_key))
-
-            existing_job = exec.job_get_by_kind_and_unique_properties(get_params)
-            if existing_job:
-                return InsertResult(existing_job, unique_skipped_as_duplicated=True)
-
-            return InsertResult(exec.job_insert(insert_params))
+    def _insert_many_exec(
+        self, exec: ExecutorProtocol, args: List[JobArgs | InsertManyParams]
+    ) -> list[InsertResult]:
+        res = exec.job_insert_many(_make_driver_insert_params_many(args))
+        return _to_insert_results(res)
 
 
-def _build_unique_get_params_and_unique_key(
+def _build_unique_key_and_bitmask(
     insert_params: JobInsertParams,
-    unique_opts: Optional[UniqueOpts],
-) -> tuple[Optional[JobGetByKindAndUniquePropertiesParam], str]:
+    unique_opts: UniqueOpts,
+) -> tuple[Optional[bytes], Optional[bytes]]:
     """
-    Builds driver get params and an advisory lock key from insert params and
-    unique options for use during a unique job insertion.
+    Builds driver get params and a unique key from insert params and unique
+    options for use during a job insertion.
     """
-
-    if unique_opts is None:
-        return (None, "")
-
     any_unique_opts = False
-    get_params = JobGetByKindAndUniquePropertiesParam(kind=insert_params.kind)
 
     unique_key = ""
 
+    if not unique_opts.exclude_kind:
+        unique_key += f"&kind={insert_params.kind}"
+
     if unique_opts.by_args:
         any_unique_opts = True
-        get_params.by_args = True
-        get_params.args = insert_params.args
-        unique_key += f"&args={insert_params.args}"
+
+        # Re-parse the args JSON for sorting and potentially filtering:
+        args_dict = json.loads(insert_params.args)
+
+        args_to_include = args_dict
+        if unique_opts.by_args is not True:
+            # Filter to include only the specified keys:
+            args_to_include = {
+                key: args_dict[key] for key in unique_opts.by_args if key in args_dict
+            }
+
+        # Serialize with sorted keys and append to unique key:
+        sorted_args = json.dumps(args_to_include, sort_keys=True)
```

Comment:
Something that just came to mind: are the JSON encodings between Go, Ruby, and Python identical on this front? Do they omit unnecessary spaces, trailing newlines, etc? Unique conflicts won't be detected properly if not.

Reply:
Hmm, yeah that might be a problem. It looks like Ruby has the same behavior as Go with minimal spacing between keys:

```
> JSON.dump({'z': 'last', 'a': 1}.sort.to_h)
=> "{\"a\":1,\"z\":\"last\"}"
```

But Python puts spacing in by default:

```
>>> json.dumps({'z': 'last', 'a': 1}, sort_keys=True)
'{"a": 1, "z": "last"}'
```

_However_, they've got a really convenient [`separators`](https://docs.python.org/3/library/json.html) keyword to `dumps` that takes the whitespace out and makes it like Ruby/Go:

```
>>> json.dumps({'z': 'last', 'a': 1}, sort_keys=True, separators=(',', ':'))
'{"a":1,"z":"last"}'
```

Reply:
Added, thanks for digging in to confirm!